File: Python/#22_generateParenthesis.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result=[]
        self.dfs(n,result,'',0,0)
        return result
    
    def dfs(self,n,result,path,left,right):
        if not self.check(left,right,n):
            return
        if len(path)==n*2:
            result.append(path)
            return
        self.dfs(n,result,path+'(',left+1,right)
        self.dfs(n,result,path+')',left,right+1)
    
    def check(self,left,right,n):
        return left>=right and left<=n and right<=n
        

File: Java/#22_generateParenthesis.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res=new ArrayList<String>();
        gp(res,0,0,n,"");
        return res;
    }
    public void gp(List<String> res, int open,int close, int n, String s){
        if(s.length()==n*2){
            res.add(s);
            return;
        }
        if(open<n){
            gp(res,open+1,close,n,s+"(");
        }if(close<open){
            gp(res,open,close+1,n,s+")");
        }
    }
}
